rq/worker.py

```diff
@@ -667,24 +669,36 @@ def perform_job(self, job, queue):
                 # to use the same exc handling when pickling fails
                 job._result = rv
 
-                self.set_current_job_id(None, pipeline=pipeline)
-
                 result_ttl = job.get_result_ttl(self.default_result_ttl)
                 if result_ttl != 0:
                     job.ended_at = utcnow()
-                    job.set_status(JobStatus.FINISHED, pipeline=pipeline)
-                    job.save(pipeline=pipeline)
 
-                    finished_job_registry = FinishedJobRegistry(job.origin,
-                                                                self.connection)
-                    finished_job_registry.add(job, result_ttl, pipeline)
+                while True:
+                    try:
+                        # if dependencies are inserted after enqueue_dependents
+                        # a WatchError is thrown by execute()
+                        pipeline.watch(job.dependents_key)
```

Comment:
It looks tempting to reduce the lines of code inside the try/except, but this does not work either. Redis' optimistic locking protects against concurrent changes between `WATCH` and `EXEC` (http://redis.io/topics/transactions). This is why `pipeline.watch(job.dependents_key)` has to come before any pipeline operations ;)

The `MULTI` command, which I've added to the `queue.enqueue_dependents`, puts the transaction again into the pipelined mode. So to be most efficient, we should leave the `watch()` + `enqueue_dependents()` at the beginning (so that's the reason I've moved it).
But I'm still convinced that we should move that code into our new `handle_successful_job_execution` ;)